### ISSUE TITLE
Some fonts are handle wrong. Wrong letters are displayed.

### ISSUE THREAD

**Issue by nfm886** [xbmc/xbmc#20130]
### Describe the bug
Here is a clear and concise description of what the problem is:

Some fonts does not works properly. Letters are swapped so for example, instead of Resize we have ResiWe or moom instead of Zoom. It is happening with Inter font and probably more fonts are affected too. Every language is affected. Also tried to change "Character set" to ISO and Windows. No change.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

Properly rendered text in UI with fonts other than Arial.

## Actual Behavior

Currently wrong letters are displayed. z is shown OK but Z is not. More letters are swapped too. Maybe depends on language choice.

## Possible Fix

I'm not sure but @vpeter4 said it might be related to this: https://github.com/xbmc/xbmc/pull/19765

### To Reproduce

Ok, so basically my skin is affected so, please follow this steps.
1. Download Cosmic skin (from another branch): https://github.com/nfm886/skin.cosmic/archive/refs/heads/kodi20-inter-font-tests.zip
2. Install it from zip file.
3. Change skin to Cosmic and see text everywhere.
Second way is to use Inter fonts.
1. Download Inter fonts: 
[fonts.zip](https://github.com/xbmc/xbmc/files/7137042/fonts.zip)
2. Use it with your skin.

### Debuglog

The debuglog can be found here:
Sorry but nothing special and related to issue. No errors at all.

### Screenshots 
Here are some links or screenshots to help explain the problem: https://imgur.com/a/feky5bR

Please note that, on 2 photos I changed font to Arial Unicode and Arial based and problems gone.

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

Basically this issue happens in whole Kodi. No matter what view you are looking right now. If letter is displayed wrong, it will be in every view.

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [X] Linux
 - [ ] OSX
 - [X] Windows
 - [ ] Windows UWP

 - Operating system version/name: CoreELEC 20 Nightly / Windows 10 latest stable
 - Kodi version: 20 Nexus

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by vpeter4** [comment]
I think the problem is because of https://github.com/xbmc/xbmc/pull/20015
Will do some more testing.

**Comment by nfm886** [comment]
> I think the problem is because of #20015.
> After reverting it Inter font works ok.
> But still it is the question if something is wrong with the font too.

Any idea for the fix? 
Here is a GH of this fotn: https://github.com/rsms/inter
Maybe there you will find some useful info :)

**Comment by vpeter4** [comment]
I don't know how to fix it.
I only know that disabling read character from cache solve the issue.
I think the issue is with the font itself :)

**Comment by nfm886** [comment]
Ok, so I should change font to make my skin compatible with K20 :eyes: 

> I only know that disabling read character from cache solve the issue.

May I know what it does and why this fix is needed?
I will test some more fonts to see if it's a common issue.
As I said before, it's a shame we can't use every fonts :(

Thank you for your interests in this topic.

**Comment by thexai** [comment]
Seems very specific issue only with one font / skin. Even that, i tried to reproduce with current master branch and works fine for me.

![screenshot00010](https://user-images.githubusercontent.com/58434170/132956981-454a7e50-45fb-4e1d-8ff4-a03d2ee016cb.png)

![screenshot00009](https://user-images.githubusercontent.com/58434170/132956985-d7e1381b-b099-4ced-b8c2-7b4a51254aae.png)

Please, provide debug log or this will not be investigated.

**Comment by nfm886** [comment]
> Seems very specific issue only with one font / skin. Even that, i tried to reproduce with current master branch and works fine for me.
> 
> ![screenshot00010](https://user-images.githubusercontent.com/58434170/132956981-454a7e50-45fb-4e1d-8ff4-a03d2ee016cb.png)
> 
> ![screenshot00009](https://user-images.githubusercontent.com/58434170/132956985-d7e1381b-b099-4ced-b8c2-7b4a51254aae.png)
> 
> Please, provide debug log or this will not be investigated.

Oh sorry. Actually I changed default font in my skin from Inter to Glory. That's why it's displayed OK on your setup...
I will update ticket for a moment. Sorry, my bad.
However as I said, log says nothing, but ok. Tomorrow I will provide log from CE20 Nightly. Hope it will help somehow.

**Comment by vpeter4** [comment]
https://pastebin.com/bDpD8EEa

![image](https://user-images.githubusercontent.com/1016324/132957719-fdc331a4-bc87-4cc3-bee2-f99adb95f298.png)

![image](https://user-images.githubusercontent.com/1016324/132957735-282ad02c-ded4-4dcf-964c-6af7ef5947eb.png)

**Comment by nfm886** [comment]
@vpeter4 thank you so much!
I will make a new branch of my skin with Inter font for tests :)

**Comment by thexai** [comment]
I can not reproduce with Estuary skin changing fonts to "Arial based" but yes with new linked Cosmic skin with inter font.

Seems it's due Inter font uses much OpenType features and ends up with glyphIndex > 255 for many ASCII chars and current cache code can't handle this. Probably this may happen with other fonts too.

![Captura de pantalla 2021-09-12 115101](https://user-images.githubusercontent.com/58434170/132983219-1f955b86-613e-4252-b120-85c1f31a9e17.png)

One possible fix may be disable caching for all symbols with  glyphIndex > 255 even if in origin are ASCII chars.

**Comment by thexai** [comment]
Test build here: https://www.mediafire.com/file/cgwwankici3wi7q/KodiSetup-20210912-c02c9287-master-x64.exe/file

**Comment by vpeter4** [comment]
> I can not reproduce with Estuary skin changing fonts to "Arial based"

Inter font must also be copied to _.kodi/media/fonts_ folder to see the effect.

> One possible fix may be disable caching for all symbols with glyphIndex > 255 even if in origin are ASCII chars.

With the above commit issue is gone. Tested with Windows version above and also self build LibreELEC.

**Comment by nfm886** [comment]
> Test build here: https://www.mediafire.com/file/cgwwankici3wi7q/KodiSetup-20210912-c02c9287-master-x64.exe/file

Just tested this build. Everything seems to work as expected indeed. Good job! :)
